### PR TITLE
Implement Rumble

### LINKED ISSUE #1659: Support for HD Rumble
#### Overview:
- Implement HD rumble for Switch controllers connected as xinput (eg. through BetterJoy)

#### Nature of Request:  

- Addition

### Why would this feature be useful?
Most games on the switch have HD rumble, so adding support for it would be ideal for playing those games.

### PR DESCRIPTION
**Pull request by shukenmg**
## Implement Rumble In Ryujinx
This PR aims to implement per-player rumble support in many games.

### Supported Controllers
- Switch Pro Controller: Works
- Joycons: Not Tested
- Xbox One Controller: Partially Works (High_Amp is weak)
- Other Controllers: Not Tested
_Please let me know if other controllers work/don't work_

### Tested Games
- ARMS: Working
- Age Of Calamity: Working
- Mario Kart 8 Deluxe: Working

**Added:**
- Super Mario Tennis: Working
- Super Mario Odyssey: Working
_Please let me know if other games work/don't work_

### Changes
Appropriate system calls are now un-stubbed.
Use of SDL2 to send vibration data to the controller.
Added SDL2 dependencies to the project.
Added rumble configuration options.
image

### Other notes
If you are using BetterJoy to connect to Ryujinx please make sure to use a version greater that 6.4b
_At this time (2020/11/04--yyyy/mm/dd) there are no versions greater than 6.4b so please use this build instead._
[BetterJoy-PreviewBuild-x64.zip](https://github.com/Ryujinx/Ryujinx/files/5485694/BetterJoy-PreviewBuild-x64.zip)
This will close #1659 

### Credits
Thanks to @anirudhb for starting this work #1436 
Kudos to @Morph1984 for his wonderful amplitude scaling formula.
PS. Do I need to add some sort of license for using the formula?

### CODE REVIEW

**Comment by Morph1984** [comment]
Hi, nice work in getting rumble working in Ryujinx. I must point out that some portions of the code were referenced from yuzu and would like to ask for proper attribution by including yuzu's copyright notice in any files that referenced yuzu's code. Once you do, we will grant you a license exemption for your usages of the code.

**Comment by shukenmg** [comment]
@Morph1984 Thanks for your help,

Which part of the code is taken from yuzu? Just the amplitude scaling?

**Comment by Thog** [comment]
@Morph1984 Please state a clear list of what part you suspect are taken from yuzu.
If your claims are right, those part would need to be rewritten as we will not make Ryujinx switch to GPL.

**Comment by shukenmg** [comment]
@Thog Thanks.
Is there an issue with using SDL2 as a dependency or is it fine to use it?

**Comment by Thog** [comment]
@shukenmg If this work correctly on Linux and Windows then yes no problem with that ^^
I plan to do a complete rewrite of all the graphics context and input management at some point so I will change that later if it's no more needed.
Also don't miss to make sure of having SDL2 present in built artifacts for Windows.

**Comment by shukenmg** [comment]
@Thog Alright, thank you very much :)

**Review by Morph1984 — COMMENTED** [review]
These are all the codes referenced from yuzu

@Thog We are making a license exemption so this code can remain so long as there is proper attribution in the form of appending yuzu's copyright notice in any files where it is used

File: Ryujinx.HLE/HOS/Services/Hid/HidDevices/NpadDevices.cs [hunk 1]
```diff
@@ -72,6 +73,20 @@ internal IEnumerable<PlayerIndex> GetSupportedPlayers()
...
+            new Dictionary<PlayerIndex, HidVibrationValue>(),
+            new Dictionary<PlayerIndex, HidVibrationValue>()
+        };
+        public Dictionary<PlayerIndex, HidVibrationValue> LastRumbleValues = new Dictionary<PlayerIndex, HidVibrationValue>();
+
+        public bool RumbleDataEqual(HidVibrationValue value1, HidVibrationValue value2)
+        {
+            return value1.AmplitudeHigh == value2.AmplitudeHigh
+                && value1.AmplitudeLow == value2.AmplitudeLow;
+        }
```
[the remark this answers is not in the document]
Comment (shukenmg) [reply Ryujinx.HLE/HOS/Services/Hid/HidDevices/NpadDevices.cs:88]: Wait really?! I coded that without looking at yuzu.

File: Ryujinx.HLE/HOS/Services/Hid/IHidServer.cs [hunk 2]
```diff
@@ -922,25 +922,35 @@ public ResultCode SendVibrationValue(ServiceCtx context)
...
+            context.RequestData.ReadInt64();
+            PlayerIndex index = (PlayerIndex)vibrationDeviceHandle;
 
+            if (context.Device.Hid.Npads.RumbleQueues.TryGetValue(index, out ConcurrentQueue<HidVibrationValue> rumbleQueue))
+            {
+                if (!context.Device.Hid.Npads.RumbleValuesBuffer[1].TryGetValue(index, out HidVibrationValue oldVibeData))
+                {
+                    oldVibeData = new HidVibrationValue();
+                }
+                if (!context.Device.Hid.Npads.RumbleDataEqual(newVibeData, oldVibeData))
```
[the remark this answers is not in the document]
Comment (shukenmg) [reply Ryujinx.HLE/HOS/Services/Hid/IHidServer.cs:944]: What does this one mean?

File: Ryujinx.HLE/HOS/Services/Hid/IHidServer.cs [hunk 3]
```diff
@@ -1013,14 +1023,41 @@ public ResultCode SendVibrationValues(ServiceCtx context)
...
+                        {
+                            context.Device.Hid.Npads.RumbleValuesBuffer[0][index] = vibrationItem;
+                            rumbleQueue.Enqueue(vibrationItem);
+                            if(rumbleQueue.Count >= 15)
+                            {
+                                _ = rumbleQueue.TryDequeue(out _);
+                            }
+                        }
+                    }
+                    else if (!context.Device.Hid.Npads.RumbleDataEqual(vibrationValues[0], oldVibeData))
```
[the remark this answers is not in the document]
Comment (shukenmg) [reply Ryujinx.HLE/HOS/Services/Hid/IHidServer.cs:1050]: What does this one mean?

File: Ryujinx/Ui/RumbleDevice.cs [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        public bool RumbleDataEqual(HidVibrationValue value1, HidVibrationValue value2)
+        {
+            return value1.AmplitudeHigh == value2.AmplitudeHigh
+                && value1.AmplitudeLow == value2.AmplitudeLow;
+        }
+
+        private static ushort ProcessAmplitude(float x)
+        {
+            // Credit to @Morph1984 for this formula
+            return (ushort)(Math.Pow(x, 0.5f) * (3.0f - 2.0f * Math.Pow(x, 015f)) * 0xFFFF);
```
[the remark this answers is not in the document]
Comment (shukenmg) [reply Ryujinx/Ui/RumbleDevice.cs:96]: Yeah... how much does this need to change to not be part of yuzu's code?

File: Ryujinx.HLE/HOS/Services/Hid/IHidServer.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (shukenmg) [reply Ryujinx.HLE/HOS/Services/Hid/IHidServer.cs:944]: nvm

File: Ryujinx.HLE/HOS/Services/Hid/IHidServer.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (shukenmg) [reply Ryujinx.HLE/HOS/Services/Hid/IHidServer.cs:1050]: nvm

File: Ryujinx.HLE/HOS/Services/Hid/HidDevices/NpadDevices.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (shukenmg) [reply Ryujinx.HLE/HOS/Services/Hid/HidDevices/NpadDevices.cs:88]: I don't think it is fair to label such a fundamental code snippet as a copy. 

Also, the way I am using this is not the same as in yuzu.

This is a function, yuzu's is an if statement with a return.

**Comment by Thog** [comment]
Looking at it again, we already have a Rumble PR already opened and with the others ongoing effort at reworking hid, this might not be right.

@Morph1984 Also about this amplification formula, this is a pretty standard formula not sure if it could really be credited.

File: Ryujinx/Ui/RumbleDevice.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (riperiperi) [reply Ryujinx/Ui/RumbleDevice.cs:96]: This isn't right anyways, in the yuzu code it is `0.15`, rather than `015`.

**Comment by shukenmg** [comment]
@Thog That other pr does not work and has many issues.

File: Ryujinx/Ui/RumbleDevice.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (shukenmg) [reply Ryujinx/Ui/RumbleDevice.cs:96]: @riperiperi Oh yeah, thanks for the catch.
It was working quite well like this though. :)

**Comment by shukenmg** [comment]
@Thog 
So is there any way I can contribute to implementing Rumble?

**Pull request closed without merging**